datafusion/optimizer/src/type_coercion.rs

```diff
@@ -735,4 +786,27 @@ mod test {
             ),
         }))
     }
+
+    #[test]
+    fn test_type_coercion_rewrite() -> Result<()> {
+        let schema = Arc::new(
+            DFSchema::new_with_metadata(
+                vec![DFField::new(None, "a", DataType::Int64, true)],
+                std::collections::HashMap::new(),
+            )
+            .unwrap(),
+        );
+        let mut rewriter = TypeCoercionRewriter::new(schema);
+        let expr = is_true(
+            lit(ScalarValue::Int32(Some(12))).eq(lit(ScalarValue::Int64(Some(13)))),
+        );
+        let expected = is_true(
+            cast(lit(ScalarValue::Int32(Some(12))), DataType::Int64)
+                .eq(lit(ScalarValue::Int64(Some(13)))),
+        );
+        let result = expr.rewrite(&mut rewriter)?;
+        assert_eq!(expected, result);
+        Ok(())
+        // TODO add more test for this
```

Comment:
I agree some more tests would be good

Reply:
https://github.com/apache/arrow-datafusion/issues/3650